Sources/MessagingPush/PushHandling/NotificationCenterFrameworkAdapter.swift

```diff
@@ -61,6 +86,47 @@ class NotificationCenterFrameworkAdapterImpl: NSObject, UNUserNotificationCenter
             completionHandler([])
         }
     }
+
+    // Swizzle method convenient when original and swizzled methods both belong to same class.
+    func swizzle(forClass: AnyClass, original: Selector, new: Selector) {
+        guard let originalMethod = class_getInstanceMethod(forClass, original) else { return }
+        guard let swizzledMethod = class_getInstanceMethod(forClass, new) else { return }
+
+        let didAddMethod = class_addMethod(forClass, original, method_getImplementation(swizzledMethod), method_getTypeEncoding(swizzledMethod))
+
+        if didAddMethod {
+            class_replaceMethod(forClass, new, method_getImplementation(originalMethod), method_getTypeEncoding(originalMethod))
+        } else {
+            method_exchangeImplementations(originalMethod, swizzledMethod)
+        }
+    }
+}
+
+// This is a bit confusing and makes the code a little more complex. However, it's the most reliable way found to get UNUserNotificationCenter.delegate swizzling to work by using an extension.
+@available(iOSApplicationExtension, unavailable)
+extension UNUserNotificationCenter {
+    // Swizzled method that gets called when `UNUserNotificationCenter.current().delegate` setter called.
+    @objc dynamic func cio_swizzled_setDelegate(delegate: UNUserNotificationCenterDelegate?) {
+        guard let diGraph = SdkInitializedUtilImpl().postInitializedData?.diGraph else {
+            // SDK not initialized yet. We hope that this function never gets called because auto push click handling will not work if this function called.
+
+            // Forward request to the original implementation implementation that we swizzled. So that the app finishes setting UNUserNotificationCenter.delegate.
+            // We have to provide the original delegate object because the SDK has not yet been initialized.
+            cio_swizzled_setDelegate(delegate: delegate)
+
+            return
+        }
+
+        diGraph.logger.debug("New UNUserNotificationCenter.delegate set. Delegate class: \(String(describing: delegate))")
+
+        diGraph.notificationCenterFrameworkAdapter.newNotificationCenterDelegateSet(delegate)
+
+        // Forward request to the original implementation that we swizzled. So that the app finishes setting UNUserNotificationCenter.delegate.
+        //
+        // Instead of providing the given 'delegate', provide CIO SDK's click handler.
+        // This will force our SDK to be the 1 push click handler of the app instead of the given 'delegate'.
+        cio_swizzled_setDelegate(delegate: diGraph.notificationCenterFrameworkAdapter.delegate)
```

Comment:
Is this method calling itself or some other method ?

Reply:
This line of code looks like recursion, but it's not. It's how you're able to call the original implementation that we overrode with swizzling. 

The best explanation that I have found on how this works is [this](https://www.uraimo.com/2015/10/23/effective-method-swizzling-with-swift/). I hope that helps. Ask more questions if you have them.